core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
+      val message = messageAndOffset.message
+      lastOffsetChecked = messageAndOffset.offset
+      // Stop searching once we see message format before 0.10.0.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (message.magic == Message.MagicValue_V0)
+        return Some(TimestampOffset(maxTimestampChecked, messageAndOffset.offset))
```

Comment:
@junrao Just want to make sure I understand the behavior we want here. It seems that sometimes we want to skip the messages w/o timestamps, and sometimes we don't skip.

For example, if we have two log segments:
segment 0: baseOffset=0, no message has timestamp, last offset 99, last modified=1000.
segment 1: baseOffset=100, the first message with a timestamp is at offset 105, and the timestamp is 2000. The second message with a timestamp is at offset 150, and the timestamp is 3000.

Now if we search for timestamp 1500, it looks that we should return offset 100, i.e. we should not ignore messages 100 - 104 that do not have timestamps, even though the timestamp we wanted is at offset 105.

On the other hand if we search for timestamp 2500, it seems that we will skip the messages 106 - 149 because they don't have timestamps.

It seems a little weird that we treat the messages w/o timestamps differently depending on where the messages are. Do you think that is the behavior we want or It would be better if we have a consistent behavior regarding messages w/o timestamps? Previously I am taking the approach that we always don't skip the messages w/o timestamps. But that does mean we may not able to return the exact message with the timestamp we wanted. What do you think?